site/journal/2020/Understanding args in Glimmer Components.md

```diff
@@ -0,0 +1,323 @@
+
+*[API]: application programming interface
+
+<aside>
+
+I’ve already written [a deep dive](https://v5.chriskrycho.com/journal/autotracking-elegant-dx-via-cutting-edge-cs/) on how the autotracking system works, but that was genuinely a *deep* dive. You don’t need to understand any of that for this post: that’s the whole point, in fact.
+
+</aside>
+
+## Part I: Understanding
+
+Let’s start by looking at [the actual Glimmer component API](https://api.emberjs.com/ember/3.23/modules/@glimmer%2Fcomponent) (using TypeScript syntax just to help see what each property is):
+
+```ts
+class Component {
+  constructor(owner: Owner, args: object) { /* ... */ }
+
+  willDestroy(): void { /* ... */ }
+
+  isDestroying: boolean;
+  isDestroyed: boolean;
+}
+```
+
+This is already a pretty small surface area, but we can trim it down to get rid of the Ember-specific pieces—
+
+- **Ember-specific:**
+    - The `owner` is passed to the constructor make things like service injections possible: **an Ember-specific feature**.
+    - The `willDestroy` hook, `isDestroying`, and `isDestroyed` are all lifecycle hooks managed by Ember.
+
+- **Non-Ember-specific:** the `args` are passed to the constructor so that `args` can be available as `this.args` on component instances. This is the piece we care about!
+
+Once we remove all of the Ember-specific pieces, here’s what we’re left with:
+
+```ts
+class Component {
+  constructor(args: object) { /* ... */ }
+}
+```
+
+The only thing this actually does is set `args` on the class. We could write that like this:
+
+```ts
+class Component {
+  args: object;
+
+  constructor(args: object) {
+    this.args = args;
+  }
+}
+```
+
+However, the `args` object on a Glimmer component is read-only: you can’t do `this.args = { evil: 'yeah!' }`. We can make it a read-only property by hiding the actual storage as a [private class field](http://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Classes/Private_class_fields) and expose only a getter to access it:
+
+```ts
+class Component {
+  #args: object;
+  get args() {
+    return this.#args;
+  }
+
+  constructor(args: object) {
+    this.#args = args;
+  }
+}
+```
+
+With that in place, only the `Component` base class can change the value of `args`: private class fields are not accessible from child classes. This is all we need to see what patterns do and don’t work with `args` in Glimmer components!
+
+Let’s imagine that we have some bucket of root state representing a user, with a name and an age. In a real app, this might correspond to data that comes from a [`Route` model hook](https://guides.emberjs.com/v3.22.0/routing/specifying-a-routes-model/), or which is stored on a service, or some other means. For our purposes, though, a simple class with some properties and methods to update them will do:
+
+```js
+class Root extends Component {
+  user = {
+    name: 'Chris Krycho';
+    age: 33;
+  };
+
+  haveABirthday() {
```

Comment:
I'd suggest renaming `haveABirthday` to something that's more obvious that you're incrementing `age` (e.g. `agedAYear` or something better because `agedAYear` makes you sound like cheese). The reason for this is I kept having to scroll up and down to see how the function was defined (since I don't have the luxury of viewing this in code with nice shortcuts). This will save a ton of searching effort from either scrolling or having duplicate split screen.

Reply:
Incorporated in a6065a6 – thanks again!